File: ios/App/Snappet/Features/Kilter/KilterGeneratorMeta.swift
import Foundation

/// The generator's sidecar metadata (`meta.json`) — the vocabulary, board-size hold masks, placement/role
/// tables and the linear grade model that drive the on-device transformer. Decoded straight from the file
/// the app lazy-downloads (`KilterGeneratorAssets`); identical to what the board-explorer web tool ships,
/// so on-device generation matches the web tool bit-for-bit. Pure value type (no ONNX, no SwiftUI) →
/// unit-tested from a small inline fixture.
struct KilterGeneratorMeta: Codable, Sendable {
    struct Specials: Codable, Sendable {
        let BOS: Int, EOS: Int, PAD: Int, MATCH: Int, NOMATCH: Int
    }
    struct Size: Codable, Sendable, Identifiable {
        let id: Int
        let name: String
        let box: [Int]   // [left, right, bottom, top]
    }
    struct Role: Codable, Sendable {
        let id: Int
        let name: String
        let color: String
    }
    /// The linear grade estimator: `bias + w_nomatch·nomatch + w_angle[angle] + Σ w_hold[token] + y_mean`.
    /// `w_hold` is indexed by **vocabulary token index** (a hold token), `angle_index` maps an angle to a
    /// `w_angle` slot. Mirrors the web tool's `st()`.
    struct GradeModel: Codable, Sendable {
        let w_hold: [Double]
        let w_angle: [Double]
        let w_nomatch: Double
        let bias: Double
        let angle_index: [String: Int]
        let y_mean: Double
        let first_hold_id: Int
    }

    /// Max sequence length the model was trained with (the input tensor is padded to this).
    let block: Int
    let pad: Int
    let specials: Specials
    let firstHoldId: Int
    /// Index → token string. Hold tokens are `HOLD_<placementId>_<roleId>`; others are
    /// `PAD/BOS/EOS/SIZE_*/ANGLE_*/GRADE_*/MATCH/NOMATCH`.
    let itos: [String]
    let sizes: [Size]
    let roles: [Role]
    /// `"<sizeId>" → [allowed token indices]` (includes EOS + that size's hold tokens). Restricting the
    /// candidate set to the mask is what makes every generated climb fit the chosen board.
    let sizeMasks: [String: [Int]]
    let gradeModel: GradeModel
    let grades: [Int]
    let gradeLabels: [String: String]
    let angles: [Int]
    let defaultSize: Int
}

/// A `KilterGeneratorMeta` plus the lookup maps the decode loop needs, precomputed once (the maps are read
/// in a hot per-token loop, so they're built up-front, not recomputed). The web tool's `tt()` prep step.
struct KilterGeneratorModel: Sendable {
    let meta: KilterGeneratorMeta
    /// token string → index.
    let stoi: [String: Int]
    /// hold-token index → placementId.
    let placementOf: [Int: Int]
    /// hold-token index → roleId.
    let roleOf: [Int: Int]
    /// roleId → role name (start/middle/finish/foot).
    let roleName: [Int: String]

    init(_ meta: KilterGeneratorMeta) {
        self.meta = meta
        var stoi = [String: Int](minimumCapacity: meta.itos.count)
        var placementOf = [Int: Int]()
        var roleOf = [Int: Int]()
        for (i, s) in meta.itos.enumerated() {
            stoi[s] = i
            if s.hasPrefix("HOLD_") {
                let parts = s.split(separator: "_")
                if parts.count == 3, let p = Int(parts[1]), let r = Int(parts[2]) {
                    placementOf[i] = p
                    roleOf[i] = r
                }
            }
        }
        self.stoi = stoi
        self.placementOf = placementOf
        self.roleOf = roleOf
        self.roleName = Dictionary(meta.roles.map { ($0.id, $0.name) }, uniquingKeysWith: { first, _ in first })
    }

    /// Decode a model from its on-disk `meta.json`.
    static func load(metaURL: URL) throws -> KilterGeneratorModel {
        let data = try Data(contentsOf: metaURL)
        return KilterGeneratorModel(try JSONDecoder().decode(KilterGeneratorMeta.self, from: data))
    }

    /// The board sizes this model supports, as the app's `KilterBoardSize` (for the Generate pickers).
    var boardSizes: [KilterBoardSize] {
        meta.sizes.map { KilterBoardSize(id: $0.id, name: $0.name, detail: "", box: nil) }
    }

    /// Grade label for a model grade index (e.g. 17 → "6a/V3"), falling back to the raw number.
    func gradeLabel(_ grade: Int) -> String { meta.gradeLabels[String(grade)] ?? String(grade) }
}

File: android/app/src/main/java/com/snappet/mobile/feature/kilter/KilterGeneratorMeta.kt
package com.snappet.mobile.feature.kilter

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * The generator's sidecar metadata (`meta.json`) — vocabulary, board-size hold masks, role table and the
 * linear grade model that drive the on-device transformer. Decoded straight from the lazy-downloaded file;
 * identical to what the board-explorer web tool ships, so on-device generation matches it. Pure value type
 * → unit-tested. Mirrors iOS `KilterGeneratorMeta`. (`ignoreUnknownKeys` skips fields we don't need, e.g.
 * `placements`/`sizeNames`.)
 */
@Serializable
data class KilterGeneratorMeta(
    val block: Int,
    val pad: Int,
    val specials: Specials,
    val firstHoldId: Int,
    /** Index → token string. Hold tokens are `HOLD_<placementId>_<roleId>`. */
    val itos: List<String>,
    val sizes: List<Size>,
    val roles: List<Role>,
    /** `"<sizeId>" → [allowed token indices]` (includes EOS + that size's hold tokens). */
    val sizeMasks: Map<String, List<Int>>,
    val gradeModel: GradeModel,
    val grades: List<Int>,
    val gradeLabels: Map<String, String>,
    val angles: List<Int>,
    val defaultSize: Int,
) {
    @Serializable
    data class Specials(val BOS: Int, val EOS: Int, val PAD: Int, val MATCH: Int, val NOMATCH: Int)

    @Serializable
    data class Size(val id: Int, val name: String, val box: List<Int>)

    @Serializable
    data class Role(val id: Int, val name: String, val color: String)

    /** Linear grade estimator `bias + w_nomatch·nomatch + w_angle[angle] + Σ w_hold[token] + y_mean`. */
    @Serializable
    data class GradeModel(
        @SerialName("w_hold") val wHold: List<Double>,
        @SerialName("w_angle") val wAngle: List<Double>,
        @SerialName("w_nomatch") val wNomatch: Double,
        val bias: Double,
        @SerialName("angle_index") val angleIndex: Map<String, Int>,
        @SerialName("y_mean") val yMean: Double,
        @SerialName("first_hold_id") val firstHoldId: Int,
    )

    companion object {
        val json = Json { ignoreUnknownKeys = true }
        fun parse(text: String): KilterGeneratorMeta = json.decodeFromString(serializer(), text)
    }
}

/**
 * A [KilterGeneratorMeta] plus the lookup maps the decode loop needs, precomputed once (read in a hot
 * per-token loop). The web tool's `tt()` prep step. Mirrors iOS `KilterGeneratorModel`.
 */
class KilterGeneratorModel(val meta: KilterGeneratorMeta) {
    /** token string → index. */
    val stoi: Map<String, Int>
    /** hold-token index → placementId. */
    val placementOf: Map<Int, Int>
    /** hold-token index → roleId. */
    val roleOf: Map<Int, Int>
    /** roleId → role name. */
    val roleName: Map<Int, String>

    init {
        val stoi = HashMap<String, Int>(meta.itos.size)
        val placementOf = HashMap<Int, Int>()
        val roleOf = HashMap<Int, Int>()
        meta.itos.forEachIndexed { i, s ->
            stoi[s] = i
            if (s.startsWith("HOLD_")) {
                val parts = s.split('_')
                if (parts.size == 3) {
                    val p = parts[1].toIntOrNull()
                    val r = parts[2].toIntOrNull()
                    if (p != null && r != null) { placementOf[i] = p; roleOf[i] = r }
                }
            }
        }
        this.stoi = stoi
        this.placementOf = placementOf
        this.roleOf = roleOf
        this.roleName = meta.roles.associate { it.id to it.name }
    }

    /** The board sizes this model supports, as the app's [KilterBoardSize] (for the Generate pickers). */
    val boardSizes: List<KilterBoardSize>
        get() = meta.sizes.map { KilterBoardSize(it.id, it.name, "", null) }

    /** Grade label for a model grade index (e.g. 17 → "6a/V3"), falling back to the raw number. */
    fun gradeLabel(grade: Int): String = meta.gradeLabels[grade.toString()] ?: grade.toString()

    companion object {
        fun parse(text: String): KilterGeneratorModel = KilterGeneratorModel(KilterGeneratorMeta.parse(text))
    }
}
